#166272340 user can create, modify, delete and read one and read all article

emmygozi opened the pull request:
#### What does this PR do?

User can create an article with images if they choose
#### Description of Task to be completed?

- user should be able to create an article
- user should be able to update an article
- user should be able to get one article
- user should be able to get all available articles
- user should be able to delete an article
- an article should have a unique slug
- an article should have a timestamp
- created migration to add a column for soft delete
- an article can have images
- an article should return the author details
- an article should return user ratings
- write a unit test for the created user and test all edge cases
- write a unit test for the updated user and test all edge cases
- write a unit test for the deleted user and test all edge cases
- write a unit test to get one user
- write a unit test to get all user
- refactor findOne and findAll articles method to a different file
- Add relation to get an article created by a user using a magic function `hasArticle`

#### How should this be manually tested?

Navigate to the endpoint:
  `POST, GET http://localhost:3000/api/v1/articles`
    `GET PUT DELETE http://localhost:3000/api/v1/articles/:slug`
#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/166272340

#### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/32358797/59530411-6b424d00-8edb-11e9-84e8-6b0af41f1c84.png)

#### Questions:

N/A

emmygozi committed:
feat-[166272340]: user create article

-user should be able to create an article
-article should have a unique slug
-article is time stamped
-article can have images

[Delivers #166272340]

despeauxz reviewed (dismissed):
Nice job, just some little comments. Please fix the hound issue, also the task specified CRUD operations for the article and need to document the endpoint on swagger. Generally, nice job

emmygozi committed:
feat-[166272340]: user create article

- fix bug in user model
- add new column image to article
- fix hound exception

[Delivers #166272340]

fidelisojeah commented on server/app.js:
```diff
@@ -21,7 +22,7 @@ app.use(bodyParser.urlencoded({
 
 app.use('/api/v1', apis);
-
+app.use('/api/v1', articlesRoutes);
```
what is this doing here exactly?

emmygozi replied:
Thank you for your observation. Fixed.

fidelisojeah commented on server/models/article.js:
```diff
@@ -1,3 +1,5 @@
+import slugCreator from '../helpers/slugCreator';
```
your path fix this
```suggestion
import slugCreator from '@helpers/slugCreator';
```

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,46 @@
+
+      const createdArticle = await Article.create(createArticleDetails);
+      const { dataValues: article } = createdArticle;
+      return res.status(201).json({ status: 'success', message: 'Article created successfully', article });
```
```suggestion
      return res.status(201).json({ status: 'success', message: 'Article created successfully', payload: article });
```

fidelisojeah reviewed (commented):
Good job

Just a few changes

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud poerations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Article should also return ratings and authors detail
- write test for models

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272340-ft-user-create-article

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,196 @@
+      const { id: userId } = req.decoded;
+      const { slug: updatedSlug } = req.params;
+
+      const user = await User.findByPk(userId);
```
make use of `req.user` here please

emmygozi replied:
Would fix that now.

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,196 @@
+      }
+
+      if (!canUpdate) {
+        return Response.error(res, 401, 'You do not have permission to update this article!');
```
this is a `403` not `401`

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -0,0 +1,196 @@
+      const deletedArticle = await Article.destroy({ where: { slug, userId } });
+
+      if (!deletedArticle) {
+        return Response.error(res, 401, 'You do not have permission to delete this article!');
```
same here `403`

fidelisojeah commented on server/helpers/articlePayload.js:
```diff
@@ -0,0 +1,43 @@
+const { User, Rating } = models;
+
+
+const articlePayload = {
```
get the updated version of this payload from @EMacco

emmygozi replied:
Syncing with @EMacco . Thanks for your kind observation.

fidelisojeah reviewed (dismissed):
Good job, just a few changes to make

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272340-ft-user-create-article

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Trim title
- Add image to article payload

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement feedback to use user instance from database in articles update

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement feedback to use user instance from database in articles update

[Delivers #166272340]
